[docs] Actions Refactoring (#4844)
DevExpress/testcafe: docs/articles/documentation/guides/basic-guides/interact-with-the-page.md

> ---

Reviewer: See the comments I left (including the last file) and apply where applicable. Write me if you have any questions.

layout: docs
title: Interact With The Page
permalink: /documentation/guides/basic-guides/interact-with-the-page.html
---
# Interact With The Page

> Test API provides a set of **actions** that enable you to interact with the webpage.

Reviewer: ```suggestion
Test API provides a set of **actions** you can use to interact with the web page.
```

> They are implemented as methods in the [test controller](../../reference/test-api/testcontroller/README.md) object. You can call them in a chained fashion.

Reviewer: "a chained fashion"? Explain

Author: @dirk-pieterse  a concept of [method chaining](https://en.wikipedia.org/wiki/Method_chaining) is referenced here.

> Below you can find the available actions with reproducible usage examples and links to their detailed descriptions.

Reviewer: ```suggestion
The available actions with reproducible usage examples and links to their detailed descriptions are listed below.
```

## Click

Click actions allow you to click an element on a page.

* [Click](../../reference/test-api/testcontroller/click.md)
* [Double-Click](../../reference/test-api/testcontroller/doubleclick.md)
* [Right-Click](../../reference/test-api/testcontroller/rightclick.md)

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Click test', async t => {
    const selectBasedOnText = Selector('label').withText('I have tried TestCafe');

    await t
        .click(selectBasedOnText);
});
```

## Press Key

> [Press Key](../../reference/test-api/testcontroller/presskey.md) action allows you to press a key or a combination of keys.

Reviewer: ```suggestion
[Press Key](../../reference/test-api/testcontroller/presskey.md) action allows you to press a key or key combination.
```

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Press Key test', async t => {
    await t
        .click('#tried-test-cafe')
        // pressing 'Space' imitates clicking the checkbox again
        .pressKey('space')
});
```

## Navigate

[Navigate](../../reference/test-api/testcontroller/navigateto.md) action navigates to the specified URL.

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Navigate To URL test', async t => {
    await t
        .navigateTo('https://github.com/DevExpress/testcafe');
});
```

## Type Text

> [Type Text](../../reference/test-api/testcontroller/typetext.md) action types the specified text into the selected input element.

Reviewer: ```suggestion
[Type Text](../../reference/test-api/testcontroller/typetext.md) action types the specified text in the selected input element.
```

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Type Text test', async t => {
    await t
        .typeText('#developer-name', 'John Doe');
});
```

## Select Text

> Actions that allow you to select text within inputs, `<textarea>` and `contentEditable` elements.

Reviewer: ```suggestion
Actions that allow you to select text in inputs, `<textarea>`, and `contentEditable` elements.
```

* [Select Text in Input Elements](../../reference/test-api/testcontroller/selecttext.md)
* [Select \<textarea\> Content](../../reference/test-api/testcontroller/selecttextareacontent.md)
* [Perform Selection within Editable Content](../../reference/test-api/testcontroller/selecteditablecontent.md)

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Select Text test', async t => {
    await t
        .typeText('#developer-name', 'John Doe')
        .selectText('#developer-name')
        .pressKey('delete');
});
```

## Hover

> [Hover](../../reference/test-api/testcontroller/hover.md) action allows to hover over the tested page.

Reviewer: ```suggestion
[Hover](../../reference/test-api/testcontroller/hover.md) action allows you to hover the mouse pointer over the tested page.
```

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://js.devexpress.com`;

test('Hover test', async t => {
    await t
        .hover('.map-container');
});
```

## Drag Elements

**Drag** actions allow you to drag elements on the tested page.

* [Drag an Element by an Offset](../../reference/test-api/testcontroller/drag.md)
* [Drag an Element onto Another One](../../reference/test-api/testcontroller/dragtoelement.md)

> > The **drag** actions do not invoke integrated browser actions such as selecting text.

Reviewer: ```suggestion
> The **drag** actions do not invoke integrated browser actions such text selection.
```

> > Use them to perform drag-and-drop actions that are processed by webpage elements, not the browser.

Reviewer: ```suggestion
> Use them to perform drag-and-drop actions that webpage elements process.
```

> To select text, use [t.selectText](../../reference/test-api/testcontroller/selecttext.md).

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://devexpress.github.io/testcafe/example/`;

test('Drag test', async t => {
    const triedCheckbox = Selector('label').withText('I have tried TestCafe');

    await t
        .click(triedCheckbox)
        .drag('#slider', 360, 0, { offsetX: 10, offsetY: 10 });
});
```

## Upload Files

> Actions that allow you to work with file upload input elements.

Reviewer: What do you mean by " to work with"?

Author: @dirk-pieterse `interact`, I'll rephrase this

* [Populate File Upload Input](../../reference/test-api/testcontroller/setfilestoupload.md)
* [Clear File Upload Input](../../reference/test-api/testcontroller/clearupload.md)

> > The file upload actions only allow you to manage the list of files to upload. These files are uploaded to the server after you initiate upload, for example, when you [click](../../reference/test-api/testcontroller/click.md) the **Upload** or **Submit** button on a webpage.

Reviewer: ```suggestion
> The file upload actions only allow you to manage the list of files you want to upload. These files are uploaded to the server after you initiate upload, for example, when you [click](../../reference/test-api/testcontroller/click.md) the **Upload** or **Submit** button on a web page.
```

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://js.devexpress.com/Demos/WidgetsGallery/Demo/FileUploader/FileSelection/jQuery/Light/`;

test('Upload Files test', async t => {
    await t
        .switchToIframe('.demo-frame')
        .setFilesToUpload('.dx-fileuploader-input', [

>             // substitute the following string with the path to a local file or multiple files you wish to upload

Reviewer: ```suggestion
            // substitute the following string with the path to a local file or multiple files you want to upload
```

            'path/to/file'
        ]);
});
```

## Take Screenshot

> Actions that allow you to take screenshots of the tested page.

Reviewer: ```suggestion
These actions allow you to take screenshots of the tested page.
```

* [Take Screenshot](../../reference/test-api/testcontroller/takescreenshot.md)
* [Take Element Screenshot](../../reference/test-api/testcontroller/takeelementscreenshot.md)

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://js.devexpress.com`;

test('Take Screenshot test', async t => {
    await t
        .takeScreenshot()
        .takeElementScreenshot('.map-container');
});
```

## Work with iframes

> At any moment, a TestCafe test has its [browsing context](https://html.spec.whatwg.org/multipage/browsers.html#windows)

Reviewer: ```suggestion
A TestCafe test's [browsing context](https://html.spec.whatwg.org/multipage/browsers.html#windows) is limited to the main window or an `<iframe>`. To use an `<iframe>` in your test,
switch the context from the main window to this `<iframe>`.
If several `<iframes>` are used in your test, you need to switch between them.
```

limited to either the main window or an `<iframe>`. To use an `<iframe>` in your test,
you need to switch the context from the main window to this `<iframe>` (and then probably back).
Likewise, if several `<iframes>` are involved in your test, you will have to switch between them.

> You can use the following methods to achieve this.

Reviewer: ```suggestion
Use the following methods to switch between windows and iframes:
```

* [Switch To \<iframe\>](../../reference/test-api/testcontroller/switchtoiframe.md)
* [Switch To Main Window](../../reference/test-api/testcontroller/switchtomainwindow.md)

**Example**

```js
import { Selector } from 'testcafe';

fixture `Example`
    .page `https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/Overview/jQuery/Light/`;

test('Working With iframe test', async t => {
    await t
        .switchToIframe('.demo-frame')
        .click('.dx-datagrid-group-panel')
        .switchToMainWindow();
});
```

## Handle Native Dialogs

> TestCafe allows you to handle native browser dialogs whenever they are invoked during the test run.

Reviewer: ```suggestion
TestCafe allows you to handle native browser dialogs that are invoked during the test run.
```

You can close [alert](https://developer.mozilla.org/en-US/docs/Web/API/Window/alert) and
[beforeunload](https://developer.mozilla.org/en-US/docs/Web/Events/beforeunload) dialogs,

> choose either option in [confirm](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm) dialogs

Reviewer: ```suggestion
choose an option in [confirm](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm) dialogs
```

> or provide text to type into [prompt](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt) dialogs.

Reviewer: ```suggestion
or provide text for [prompt](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt) dialogs.
```

* [Set Native Dialog Handler](../../reference/test-api/testcontroller/setnativedialoghandler.md)
* [Get Native Dialog History](../../reference/test-api/testcontroller/getnativedialoghistory.md)

## Resize Window

> Resize window actions allow you to maximize a browser window or resize it to fit specified width and height or a type of device.

Reviewer: ```suggestion
Use resize window actions to maximize a browser window or resize it to fit a specified width and height or type of device.
```
